File: maramazza/py/4.py
def createDictionaryFromEntry(entry):
    result = dict()

    entry = entry.replace(' ', '\n')
    entry = entry.split('\n')

    for info in entry:
        result[info[0:3]] = info[4:]

    return result


if __name__ == "__main__":
    data = []

    with open('./../data/4.txt', 'r') as f:
        data = f.read().split('\n\n')

    # part 1
    totPart1 = 0

    requiredFields = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid']

    for entry in data:
        required = 0
        for field in requiredFields:
            if field in entry:
                required += 1
        if required >= len(requiredFields):
            totPart1 += 1

    print(f'Result Part 1: {totPart1}')

    # part 2
    totPart2 = 0

    for entry in data:
        entryDictionary = createDictionaryFromEntry(entry)

        valid = 0

        try:
            # byr
            byr = int(entryDictionary['byr'])
            if byr >= 1920 and byr <= 2002:
                valid += 1

            # iyr
            iyr = int(entryDictionary['iyr'])
            if iyr >= 2010 and iyr <= 2020:
                valid += 1

            # eyr
            eyr = int(entryDictionary['eyr'])
            if eyr >= 2020 and eyr <= 2030:
                valid += 1

            # hgt
            hgt = entryDictionary['hgt']

            if hgt.endswith('cm'):
                if int(hgt[:-2]) >= 150 and int(hgt[:-2]) <= 193:
                    valid += 1
            elif hgt.endswith('in'):
                if int(hgt[:-2]) >= 59 and int(hgt[:-2]) <= 76:
                    valid += 1
            else:
                pass

            # hcl
            hcl = entryDictionary['hcl']
            if hcl.startswith('#') and len(hcl) == 7:
                valid += 1
            int(hcl[1:], 16)

            #ecl
            ecl = entryDictionary['ecl']
            if ecl == 'amb' or ecl == 'blu' or ecl == 'brn' or ecl == 'gry' or \
               ecl == 'grn' or ecl == 'hzl' or ecl == 'oth':
                valid += 1

            # pid
            pid = entryDictionary['pid']
            if pid.isdigit() and len(pid) == 9:
                valid += 1
        except:
            valid = 0

        if valid >= 7:
            totPart2 += 1

    print(f'Result Part 2: {totPart2}')

File: maramazza/php/4.php
<?php
$res1 = 0;
$resOK = 0;
$rules = array('byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid', 'cid');
/*

byr (Birth Year) - four digits; at least 1920 and at most 2002.
iyr (Issue Year) - four digits; at least 2010 and at most 2020.
eyr (Expiration Year) - four digits; at least 2020 and at most 2030.
hgt (Height) - a number followed by either cm or in:
If cm, the number must be at least 150 and at most 193.
If in, the number must be at least 59 and at most 76.
hcl (Hair Color) - a # followed by exactly six characters 0-9 or a-f.
ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth.
pid (Passport ID) - a nine-digit number, including leading zeroes.
cid (Country ID) - ignored, missing or not.

*/

function checkValue($field, $value)
{
    switch ($field) {
        case 'byr':
            if (strlen($value) == 4) {
                if ((int)$value >= 1920 && (int)$value <= 2002)
                    return true;
                else
                    return false;
            } else {
                return false;
            }
            break;
        case 'iyr':
            if (strlen($value) == 4) {
                if ((int)$value >= 2010 && (int)$value <= 2020)
                    return true;
                else
                    return false;
            } else {
                return false;
            }
            break;
        case 'eyr':
            if (strlen($value) == 4) {
                if ((int)$value >= 2020 && (int)$value <= 2030)
                    return true;
                else
                    return false;
            } else {
                return false;
            }
            break;
        case 'hgt':
            break;
        case 'hcl':
            break;
        case 'ecl':
            break;
        case 'pid':
            break;
    }
}

$content = file_get_contents('./../data/4.txt');
$rows = preg_split("#\n\s*\n#Uis", $content);

$passports = array();
foreach ($rows as $row) {
    $passports[] = preg_split("/ /", str_replace("\n", " ", $row));
}
$tot = count($passports);
$check = array();
foreach ($passports as $key => $passport) {
    $elems = count($passport);
    for ($i = 0; $i < $elems; $i++) {
        $check[$key][] = substr($passport[$i], 0, 3);
    }
    $diff = array_diff($rules, $check[$key]);
    if (!$diff or ((count($diff) == 1) && (in_array('cid', $diff)))) {
        $res1++;
        print_r($passport);
    }
}
echo "res1: " . $res1;
